---
title: "RbashGEO"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

###The purpose of the package RbashGEO
- Help users to populate the NGS raw data processing commands in bash with little code.

- Users only need to provide a **collumn data**, that means a table record the design and names of the fastq files, they could be the **SRR RUN** ids on GEO.

###In stall it with the command below

```{r}
devtools::install_github("ZhenWei10/RbashGEO")
```

- The functions of this package can help users to:
1. Download and decompress raw fastq data from NCBI.
2. QC, trim, and align the fastq files with popular RNA-Seq command line tools.
3. Count the alignment results with a user provided annotation by `GRanges` in R, the count is conducted by `SummarizeOverlap` function in `GenomicAlignment` package, which is transplanted from the **HTSeq-count** in python.

###A standardized work flow 

- First we need to prepare a `Coldata`.
```{r}
library(RbashGEO)
Coldata_example
```

Then we align them
```{r}
library(RbashGEO)
library(dplyr)
Coldata_new <- read.csv("./Coldata_M14new.csv")

mapply(
 function(x,y){
   Rhisat2(Fastq_file_name = x,
           Paired = y,
           parallel_num = 1,
           Fastq_directory = getwd()) %>% Rnohup(.,paste0(x,"_hisat2"))}, 
 Coldata_new$SRR_RUN,
 (Coldata_new$Lib == "Paired")
)
```

