---
title: "RbashGEO"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

###The purpose of the package RbashGEO
- Help users to populate the RNA-Seq raw data processing commands in bash with little code, i.e. it can make you life easier if you frequently dowload and align RNA-Seq data from [**GEO**](https://www.ncbi.nlm.nih.gov/geo/).

- Users only need to provide a **collumn data**, that means a table recording the design and names of the fastq files; The sample ids could directly be the **RUN IDs** from GEO.

###Install it with the command below

```{r,eval=FALSE}
devtools::install_github("ZhenWei10/RbashGEO")
```

- The functions of this package can help users to:
1. Download and decompress many raw fastq data from GEO easily.
2. QC, trim, and align the fastq files with popular RNA-Seq command line tools, but with well organized output in R.
3. Count the alignment results with a user provided annotation by `GRanges` in R, the count is conducted by `SummarizeOverlap` function in `GenomicAlignment` package, which is transplanted from the [**HTSeq-count**](http://htseq.readthedocs.io/en/release_0.9.1/) in python.
4. The most significant utility of this package is to reduce the repetitive bash coding and in the mean time generate compact results for downstream analysis.

###A template work flow for RNA-Seq

-  First, we need to prepare a **collumn data**, such as the example below.
```{r}
library(RbashGEO)
knitr::kable( Coldata_example[1:6,1:6] )
```

Only 2 collumns: `SRR_RUN` and `Lib` are neccessary to complete the work flow of this package, but addition collumns are valuable for their own good. 

The design of the collumn table is summarized from [this page](https://www.ncbi.nlm.nih.gov/Traces/study/?acc=SRP103072) on GEO.


- Then, we should download them from GEO and run fastqc for quality control:


```{r,eval=FALSE}
library(RbashGEO)
mapply(function(x,y) Rnohup(WgetQC(x,y),x), Coldata_example$SRR_RUN,Coldata_example$Lib == "Paired")
```


-  Next, we should align them with [**hisat2**](https://ccb.jhu.edu/software/hisat2/index.shtml) (you should first check the fastqc reports, if they are not OK, you should use `Rtrim_galore` before alignment.):

```{r,eval = FALSE}
library(RbashGEO)
library(dplyr)

mapply(
 function(x,y){
   Rhisat2(Fastq_file_name = x,
           Paired = y,
           parallel_num = 1,
           Fastq_directory = getwd()) %>% Rnohup(.,paste0(x,"_hisat2"))}, 
 Coldata_example$SRR_RUN,
 (Coldata_example$Lib == "Paired")
)
```


-  Check the organized alignment reports in R all together with this command.

```{r,out.width = 30}
hisat2_report <- RbashGEO::Check_hisat2_reports()
knitr::kable(hisat2_report[1:6,-1*c(2,3,4)])
```

You should see a `data.frame` with summarized outputs of hisat2.


- Then, convert the sam into bam with some desired filters on [SAM flags](https://broadinstitute.github.io/picard/explain-flags.html).

```{r,eval=FALSE}
library(RbashGEO)

sapply(
  Coldata_example$SRR_RUN,
  function(x) Rnohup(
  Rsamtools_view(x,sam_end = ".sam",parallel_num = 1,flag_filter = 2820),
  x)
)
```

- Finally, count the reads. The reads count are stored with `SummarizedExperiment` object in R.

At this step, we need to provide a `GRanges` object for annotation, in this case we use `Annotation_gr` as an example.

```{r,eval=FALSE}
Count_SRRs(Coldata_example$SRR_RUN,"./",Annotation_gr,"Example_human_SE")
```

At last, the obtained `SummarizedExperiment` object could be conveniently handeled with other QC, inference, and learning work flows.

If you are interested in MeRIP data, you could use [meripQC](https://github.com/ZhenWei10/meripQC); 

if you are interested in the analysis of the RNA modification data, you could look [m6ALogisticModel](https://github.com/ZhenWei10/m6ALogisticModel) for more information.

```{r}
sessionInfo()
```

